src/QsCompiler/SyntaxProcessor/TypeInference/Constraint.fs

```diff
@@ -7,52 +7,81 @@ open Microsoft.Quantum.QsCompiler.SyntaxTokens
 open Microsoft.Quantum.QsCompiler.SyntaxTree
 open Microsoft.Quantum.QsCompiler.Transformations.QsCodeOutput
 
-type internal Constraint =
-    | Adjointable
-    | Callable of input: ResolvedType * output: ResolvedType
-    | CanGenerateFunctors of functors: QsFunctor Set
-    | Controllable of controlled: ResolvedType
-    | Equatable
-    | HasPartialApplication of missing: ResolvedType * result: ResolvedType
-    | Indexed of index: ResolvedType * item: ResolvedType
-    | Integral
-    | Iterable of item: ResolvedType
-    | Numeric
-    | Semigroup
-    | Wrapped of item: ResolvedType
-
-module internal Constraint =
-    /// Pretty prints a type.
-    let private prettyType: ResolvedType -> _ = SyntaxTreeToQsharp.Default.ToCode
+type ClassConstraint =
+    | Adjointable of operation: ResolvedType
+    | Callable of callable: ResolvedType * input: ResolvedType * output: ResolvedType
+    | Controllable of operation: ResolvedType * controlled: ResolvedType
+    | Eq of ResolvedType
```

Comment:
nit: Is there a particular reason why this was shortened to `Eq`? `Equatable` even though is more verbose, seems to me less readable. Same with `PartialAp`.

Reply:
On a related note, is there a reason to remove the Can/Has prefixes?

Reply:
Same question for `Wrapped` to `Unwrap`, was there any particular reason for the name change?

Reply:
Eq: Mostly just for alignment with other equality type classes, like [Haskell's Eq](https://hackage.haskell.org/package/base-4.16.1.0/docs/Prelude.html#t:Eq) and [Rust's Eq](https://doc.rust-lang.org/std/cmp/trait.Eq.html).

Unwrap: I think that the old name Wrapped was a little misleading; you can have a type that "wraps" another type (e.g. `option`) but which doesn't have a total unwrap function (e.g. `Option.get` is partial). The Unwrap class is for types that support Q#'s unwrap operator `!` which is total. So they are defined by the ability to unwrap, not by the fact that they are wrapping.

PartialAp: No strong reason, I renamed this back to HasPartialApplication:

(Can)GenerateFunctors: This one is just kind of hard to come up with a name for... I renamed it to HasFunctorsIfOperation, since I think that's most accurate - it only applies to operation types, other types trivially satisfy the constraint even if they don't have the functors.

Reply:
Thank you for the explanation!